src/ProgrammableDevice.cc

```diff
@@ -0,0 +1,109 @@
+#include "ProgrammableDevice.hh"
+
+#include "CommandController.hh"
+#include "MSXCliComm.hh"
+#include "MSXCPUInterface.hh"
+#include "MSXException.hh"
+#include "StringSetting.hh"
+#include "serialize.hh"
+
+#include "narrow.hh"
+
+namespace openmsx {
+
+ProgrammableDevice::ProgrammableDevice(const DeviceConfig& config)
+	: MSXDevice(config)
+	, portListSetting(
+		getCommandController(), tmpStrCat(getName(), " ports"),
+		"List of ports are added to the pool of IO ports this extension scans for", "")
+	, resetCallback(getCommandController(),
+		getName() + " reset callback",
+		"Tcl proc to call when reset was detected, no parameters and return values are "
+		"used", "", Setting::Save::NO)
+	, readCallback(getCommandController(),
+		getName() + " read callback",
+		"Defines a Tcl proc that receives a port number (byte) and returns a byte when "
+		"the MSX is reading a port from the device", "", Setting::Save::NO)
+	, writeCallback(getCommandController(),
+		getName() + " write callback",
+		"Defines a Tcl proc that receives a port number (byte) and a byte when the "
+		"MSX is writing a port from the device", "", Setting::Save::NO)
+{
+	portListSetting.attach(*this);
+	try {
+		outPorts = inPorts = parseAllPorts(portListSetting);
+	} catch (MSXException&) {
+		// silently ignore error in settings.xml
+		portListSetting.setString("");
```

Comment:
This only happens when a saved setting value is already wrong. Or when the user entered this setting, and only later inserted the `ProgrammableDevice` (in a running MSX instance).

Genuine question, i don't know the answer: is it better to silently clear the setting, or leave the current wrong value (maybe give a warning), so that the user can still query and correct the value?

Reply:
> Genuine question, i don't know the answer: is it better to silently clear the setting, or leave the current wrong value (maybe give a warning), so that the user can still query and correct the value?

If we leave it there, the command
```set {Programmable Device ports}```
will still display the wrong data, but OpenMSX will not use it, which I guess could be confusing for the user.